src/channel/fs_prover_channel.rs

```diff
@@ -0,0 +1,252 @@
+use crate::channel::pow::ProofOfWorkProver;
+use crate::channel::{Channel, ChannelStates, FSChannel, ProverChannel};
+use crate::randomness::prng::Prng;
+use ark_ff::{BigInteger, PrimeField};
+use num_bigint::BigUint;
+use sha3::digest::{Digest, Output};
+use std::marker::PhantomData;
+
+pub struct FSProverChannel<F: PrimeField, D: Digest, P: Prng> {
+    pub _ph: PhantomData<(F, D)>,
+    pub prng: P,
+    pub proof: Vec<u8>,
+    pub states: ChannelStates,
+}
+
+impl<F: PrimeField, D: Digest, P: Prng> FSProverChannel<F, D, P> {
+    pub fn new(prng: P) -> Self {
+        Self {
+            _ph: PhantomData,
+            prng,
+            proof: vec![],
+            states: Default::default(),
+        }
+    }
+}
+
+impl<F: PrimeField, D: Digest, P: Prng> Channel for FSProverChannel<F, D, P> {
+    type Field = F;
+
+    fn draw_number(&mut self, upper_bound: u64) -> u64 {
+        assert!(
+            !self.states.is_query_phase(),
+            "Prover can't receive randomness after query phase has begun."
+        );
+
+        let raw_bytes = self.draw_bytes(std::mem::size_of::<u64>());
+        let number = u64::from_be_bytes(raw_bytes.try_into().unwrap());
+
+        assert!(
+            upper_bound < 0x0001_0000_0000_0000,
+            "Random number with too high an upper bound"
+        );
+
+        number % upper_bound
+    }
+
+    fn draw_felem(&mut self) -> Self::Field {
+        assert!(
+            !self.states.is_query_phase(),
+            "Prover can't receive randomness after query phase has begun."
+        );
+
+        // from stone-prover/src/starkware/algebra/fields/big_prime_constants.h
+        //   static constexpr ValueType kMaxDivisible = 0xf8000000000000000000000000000000000000000007c000000000000000001f_Z;
+        let max_divisible = BigUint::parse_bytes(
+            b"f8000000000000000000000000000000000000000007c000000000000000001f",
+            16,
+        )
+        .unwrap();
+        let mod_felt252 = BigUint::parse_bytes(
+            b"800000000000000000000000000000000000000000040000000000000000001",
+            16,
+        )
+        .unwrap();
+
+        let mut raw_bytes: Vec<u8>;
+        let mut random_biguint: BigUint;
+        loop {
+            raw_bytes = self.draw_bytes(((Self::Field::MODULUS_BIT_SIZE + 7) / 8) as usize);
+            random_biguint = BigUint::from_bytes_be(&raw_bytes);
+            if random_biguint < max_divisible {
+                random_biguint = random_biguint.modpow(&BigUint::from(1u64), &mod_felt252);
+                break;
+            }
+        }
+
+        let field_element: F = Self::Field::from_bigint(
+            <Self::Field as PrimeField>::BigInt::try_from(random_biguint).unwrap(),
+        )
```

Comment:
@rot256 
I had to use BigUint to divide some large integers (smaller than max_divisible) by mod_felt252. 
The current code is messy and lacks generality. 

I've been trying to find a way to calculate max_divisible once and assign it to a static constant, but this has been time-consuming. For now, I've left the implementation as is.

Reply:
Also, I cannot use `Self::Field::from_be_bytes_mod_order()`, output differs 😓

Reply:
```rust
impl<F: PrimeField, D: Digest, P: Prng> FSProverChannel<F, D, P> {
    fn modulus() -> &'static BigUint {
        static MODULUS: OnceLock<BigUint> = OnceLock::new();
        MODULUS.get_or_init(|| F::MODULUS.into())
    }

    fn max_divislble() -> &'static BigUint {
        static MAX_VALUE: OnceLock<BigUint> = OnceLock::new();
        MAX_VALUE.get_or_init(|| {
            let modulus = F::MODULUS.into();
            let size: usize = ((F::MODULUS_BIT_SIZE + 7) / 8) as usize;
            let max = BigUint::from_bytes_be(&vec![0xff; size]);
            let quotient = max.div(&modulus);
            quotient * modulus
        })
    }
}
```

Removed non-generic constant.